### PR TITLE
fix(normalize): refuse a canonicalization that would change the sequence

### PR DESCRIPTION
**Pull request by nh13**
> Stacked on #1239 → #1238 → #1237 → #1236. Base retargets as those merge.

## Why

The sequence-first canonicalizer verified its own output with a `debug_assert_eq!` — which is compiled out of exactly the release builds that process real data. The whole point of #1234 is that a normalizer silently emitting a **different variant** is the worst failure this crate can have, so the check belongs at runtime.

This is not hypothetical: that assertion **caught a real bug during development**, where 3'-shifting a `delins` piece changed the resulting sequence because `shuffle`'s rotation invariant only holds for pure deletions and insertions. Had the canonicalizer shipped with the check debug-only, that class of defect would reach users silently.

## What changes

Re-derive the resulting sequence from the rebuilt members and compare. If it does not round-trip, **fall back to the per-member pipeline's output** rather than emitting the re-derivation — a missed canonicalization is a cosmetic defect; a changed sequence is data corruption.

Cost is one window-sized apply on a path that has already fetched and applied that window once.

8797/8797 pass; `fmt` and `clippy` clean.

## Summary by CodeRabbit

* **Bug Fixes**
  * Added a runtime consistency check during sequence canonicalization.
  * Prevents silent sequence changes by falling back to the safer processing path when results do not match.

### CODE REVIEW

**Comment by nh13** [comment]
Converted to draft only because its parent #1237 is blocked — **the change itself reviewed well** and is a genuine correctness win.

Independent verification: 8000 fuzzed cis alleles run against this commit and its parent, with every differing output adjudicated by applying it back to the reference.

| outcome | count |
|---|---|
| identical | 7586 |
| **parent changed the sequence, guard correct** | **389** |
| parent correct, guard wrong | **0** |
| both wrong | 11 |
| guard refused although both were sequence-correct | 12 |

A debug build of the parent panics on the first differing case. So this is a real ~5%-of-random-cis-alleles corruption that release builds were silently shipping. Polarity is correct (`None` cannot read as success), the check cannot pass while anything in the window differs, and it is **net faster** than the parent (0.11 s vs 0.13 s on the fuzz) because refusing early skips downstream work.

Three things to fix before un-drafting:

1. **Keep the `debug_assert` alongside the refusal, don't replace it.** The refusal makes the output safe but makes the underlying canonicalizer defect permanently invisible — it fires on ~5% of random cis alleles and **zero** of the 4929 existing tests, so once merged nothing can ever fail because of it again. `debug_assert!(round_trips, …)` then `if !round_trips { return variants; }`.
2. **The `None` branch conflates "changed the sequence" with "unrepresentable", masking a distinct bug.** The 12 refusals where both outputs were sequence-correct are cases where `anchor_for_piece` produced a pure-insertion piece at window offset 0, so `build_naedit` rendered **`g.0_1insA`** — genomic position 0, not valid HGVS. Reachable whenever `w_lo` clamps to 1 and the trimmed block starts at offset 0. Refusing is the right behaviour; folding it into the sequence check is what hides it. Separate the two arms.
3. **Add tests for the refusal path** — currently none. Two concrete reproducers are in the review: a sequence-changing case (`g.[20C>G;22_23dup]` on a given reference) and the `None` case (`g.[1_4delinsATC;6delinsAAT]`). Cheaper still: extract the predicate into a named function and unit-test all three arms, which pins the polarity that `is_some_and` makes easy to invert.

Also noted: "fall back to the per-member pipeline" is not a guaranteed-safe destination — 11/8000 are wrong on both branches, and the fallback itself can emit overlapping members (`g.[3_4dup;4_5inv]`). Pre-existing in the per-member `dup` shift, but this commit now depends on that path; worth a linked follow-up.

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai pause

**Comment by nh13** [comment]
@coderabbitai full review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
Consolidated into #1237 and closing.

This PR replaced a `debug_assert_eq!` that #1237 itself adds, so it was a fixup to that PR rather than an independent change. CodeRabbit flagged the same `debug_assert` **on #1237**, and the answer there had to be "see #1240" — a review round spent on a split that bought nothing. Its commit is now the third of three on `fix/issue-1235-sequence-first-canonicalization`, with its review finding applied.

The finding was right: collapsing the round trip into one boolean threw away the distinction between "cannot tell" and "the canonicalizer is wrong". The three outcomes are now separated —

- `collect_canonical_edits` returning `None` — the rebuilt form will not lower back to edits;
- `apply_edits_to_window` returning `None` — those edits will not re-apply;

both are ordinary "decline and let the per-member pipeline answer". The third is not: a round trip that *succeeds* and disagrees means this module built a form denoting different bases, which is a defect here rather than an input it cannot serve. That case now carries its own `debug_assert_eq!` — loud in tests and development — with the runtime refusal still carrying release, which was the point of the PR.

File: src/normalize/merge.rs [hunk 1]
```diff
@@ -2042,13 +2042,23 @@ pub(crate) fn canonicalize_from_sequence<P: ReferenceProvider>(
...
+    //
+    // If it does not round-trip, fall back to the per-member pipeline's output
+    // rather than emitting the re-derivation. A missed canonicalization is a
+    // cosmetic defect; a changed sequence is data corruption.
+    let round_trips = collect_canonical_edits(&rebuilt, kind, body, &template_accession)
+        .and_then(|edits| apply_edits_to_window(&edits, &ref_bytes, w_lo))
+        .is_some_and(|reapplied| reapplied == result);
+    if !round_trips {
+        return variants;
+    }
```
[the remark this answers is not in the document]
Comment (nh13) [reply src/normalize/merge.rs:2061]: Fixed in #1237 (this commit moved there), taking the suggested shape.

The three outcomes are now separate. `collect_canonical_edits` returning `None` (the rebuilt form will not lower back to edits) and `apply_edits_to_window` returning `None` (those edits will not re-apply) are both ordinary "cannot tell" — decline, and the per-member pipeline's output stands. A round trip that *succeeds* and disagrees is different in kind: it means this module built a form denoting different bases, which is a defect here rather than an input it cannot serve. That case keeps its own `debug_assert_eq!` with the original diagnostic message, and the runtime refusal still carries release — which was the point of the PR.
